File: Array/fanbei/0625/java/Solution06_25.java
package com.example.lsn37_dagger2_subcomponent;

/**
 * 环路检测
 */
public class Solution06_25 {

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public static class Solution {
        public ListNode detectCycle(ListNode head) {
            if (head == null || head.next == null) {
                return null;
            }
            ListNode fast = head;
            ListNode slow = head;
            while (fast != null && fast.next != null) {
                fast = fast.next.next;
                slow = slow.next;
                if (fast == slow) {
                    break;
                }
            }
            if (fast != slow) {
                return null;
            }
            fast = head;
            while (fast != slow) {
                fast = fast.next;
                slow = slow.next;
            }
            return fast;
        }
    }
}

File: Array/fanbei/0625/c/solution06_25.c
#include <stdio.h>
#include <stdlib.h>

/**
 * 环路检测
 */

struct ListNode{
    int val;
    struct ListNode *next;
};

struct ListNode* detectCycle(struct ListNode* head){
    if(!head||head->next){
        return NULL;
    }
    struct ListNode* fast = head;
    struct ListNode* slow = head;
    while (fast && fast->next){
        slow = slow->next->next;
        fast = fast->next;
        if(fast = slow){
            break;
        }
    }
    if(fast!=slow){
        return NULL;
    }
    fast = head;
    while (fast!=slow){
        fast =fast->next;
        slow = slow->next;
    }
    return fast;
}

